src/NCalc/Expression.cs

```diff
@@ -88,107 +89,75 @@ public static bool CacheEnabled
         /// </summary>
         private static void CleanCache()
         {
-            try
+            foreach (var kvp in _compiledExpressions)
             {
-                var keysToRemove = new List<string>();
-                foreach (var de in _compiledExpressions)
-                {
-                    if (!de.Value.IsAlive)
-                    {
-                        keysToRemove.Add(de.Key);
-                    }
-                }
-
-                foreach (string key in keysToRemove)
+                if (!kvp.Value.TryGetTarget(out _))
                 {
-                    _compiledExpressions.TryRemove(key, out _);
+                    _compiledExpressions.TryRemove(kvp.Key, out _);
                     //Debug.WriteLine("Cache entry released: " + key);
                 }
             }
-            finally
-            {
-            }
         }
 
         #endregion
 
         public static LogicalExpression Compile(string expression, bool nocache)
         {
-            LogicalExpression logicalExpression = null;
-
             if (_cacheEnabled && !nocache)
             {
-                if (_compiledExpressions.ContainsKey(expression))
+                if (_compiledExpressions.TryGetValue(expression, out var wr))
                 {
                     //Debug.WriteLine("Expression retrieved from cache: " + expression);
 
-                    var wr = _compiledExpressions[expression];
-                    logicalExpression = wr.Target as LogicalExpression;
-
-                    if (wr.IsAlive && logicalExpression != null)
-                    {
-                        return logicalExpression;
-                    }
+                    if (wr.TryGetTarget(out var target))

```

Comment:
The property IsAlive is probably lighter weight than TryGetTarget and could be used to short-circuit, but perhaps the chance of a garbage collected expression is very low and this adds overhead.

`if (wr.IsActive && wr.TryGetTarget(out var target))`

Reply:
Sorry, `WeakReference` contains only` IsAlive` property, but there there is a [remark](https://learn.microsoft.com/en-us/dotnet/api/system.weakreference.isalive?view=net-6.0#remarks):

> Because an object could potentially be reclaimed for garbage collection immediately after the [IsAlive](https://learn.microsoft.com/en-us/dotnet/api/system.weakreference.isalive?view=net-6.0) property returns true, using this property is not recommended unless you are testing only for a false return value.